src/prof_recent.c

```diff
@@ -0,0 +1,522 @@
+		/* If upper limit is reached, simply shift the dummy. */
+		assert(prof_recent_alloc_max_get(tsd) != -1);
+		assert(!prof_recent_alloc_is_empty(tsd));
+		prof_recent_alloc_dummy = prof_recent_alloc_dummy->next;
+		old_alloc_tctx = prof_recent_alloc_dummy->alloc_tctx;
+		assert(old_alloc_tctx != NULL);
+		old_dalloc_tctx = prof_recent_alloc_dummy->dalloc_tctx;
+		prof_recent_alloc_evict_edata(tsd, prof_recent_alloc_dummy);
+	} else {
+		/* Otherwise use the new node as the dummy. */
+		assert(prof_recent_alloc_max_get(tsd) == -1 ||
+		    prof_recent_alloc_count < prof_recent_alloc_max_get(tsd));
+		if (reserve == NULL) {
+			goto label_rollback;
+		}
+		reserve->next = prof_recent_alloc_dummy->next;
+		prof_recent_alloc_dummy->next = reserve;
+		prof_recent_alloc_dummy = reserve;
+		reserve = NULL;
+		old_alloc_tctx = NULL;
+		old_dalloc_tctx = NULL;
+		++prof_recent_alloc_count;
+	}
+
+	assert(prof_recent_alloc_max_get(tsd) == -1 ||
+	    (prof_recent_alloc_max_get(tsd) > 0 &&
+	    prof_recent_alloc_count <= prof_recent_alloc_max_get(tsd)));
+	assert(!prof_recent_alloc_is_empty(tsd));
+
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (reserve != NULL) {
+		idalloctm(tsd_tsdn(tsd), reserve, NULL, NULL, true, true);
+	}
+
+	/*
+	 * Asynchronously handle the tctx of the old node, so that there's no
+	 * simultaneous holdings of prof_recent_alloc_mtx and tdata->lock.
+	 * In the worst case this may delay the tctx release but it's better
+	 * than holding prof_recent_alloc_mtx for longer.
+	 */
+	if (old_alloc_tctx != NULL) {
+		decrement_recent_count(tsd, old_alloc_tctx);
+	}
+	if (old_dalloc_tctx != NULL) {
+		decrement_recent_count(tsd, old_dalloc_tctx);
+	}
+	return;
+
+label_rollback:
+	assert(edata_prof_recent_alloc_get(tsd, edata) == NULL);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (reserve != NULL) {
+		idalloctm(tsd_tsdn(tsd), reserve, NULL, NULL, true, true);
+	}
+	decrement_recent_count(tsd, tctx);
+}
+
+ssize_t
+prof_recent_alloc_max_ctl_read() {
+	/* Don't bother to acquire the lock. */
+	return prof_recent_alloc_max_get_no_lock();
+}
+
+ssize_t
+prof_recent_alloc_max_ctl_write(tsd_t *tsd, ssize_t max) {
+	assert(max >= -1);
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	const ssize_t old_max = prof_recent_alloc_max_update(tsd, max);
+
+	if (max == -1 || prof_recent_alloc_count <= max) {
+		/* Easy case - no need to alter the list. */
+		malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		return old_max;
+	}
+
+	prof_recent_t *begin = prof_recent_alloc_dummy->next;
+	do {
+		assert(!prof_recent_alloc_is_empty(tsd));
+		prof_recent_t *node = prof_recent_alloc_dummy->next;
+		assert(node != prof_recent_alloc_dummy);
+		prof_recent_alloc_evict_edata(tsd, node);
+		prof_recent_alloc_dummy->next = node->next;
+		--prof_recent_alloc_count;
+	} while (prof_recent_alloc_count > max);
+	prof_recent_t *end = prof_recent_alloc_dummy->next;
+	assert(begin != end);
+
```

Comment:
nit: add a `if (config_debug) ...` to walk through the list and count total == prof_recent_alloc_count.

Reply:
Extracted that as a common assertion utility function and constantly call it to make sure the list is kept in good shape.